modules/apps/object/object-service/src/main/java/com/liferay/object/service/impl/ObjectValidationRuleSettingLocalServiceImpl.java

```diff
@@ -53,6 +65,159 @@ public ObjectValidationRuleSetting addObjectValidationRuleSetting(
 			objectValidationRuleSetting);
 	}
 
+	@Override
+	public List<ObjectValidationRuleSetting> updateObjectValidationRuleSettings(
+			ObjectValidationRule objectValidationRule,
+			List<ObjectValidationRuleSetting> objectValidationRuleSettings)
+		throws PortalException {
+
+		List<ObjectValidationRuleSetting> keyObjectValidationRuleSettings =
+			new ArrayList<>();
+
+		List<ObjectValidationRuleSetting> outputObjectValidationRuleSettings =
+			new ArrayList<>();
+
+		for (ObjectValidationRuleSetting objectValidationRuleSetting :
+				objectValidationRuleSettings) {
+
+			if (objectValidationRuleSetting.compareName(
+					ObjectValidationRuleSettingConstants.
+						NAME_COMPOSITE_KEY_OBJECT_FIELD_ID)) {
+
+				keyObjectValidationRuleSettings.add(
+					objectValidationRuleSetting);
+
+				continue;
+			}
+
+			outputObjectValidationRuleSettings.add(objectValidationRuleSetting);
+		}
+
+		long userId = objectValidationRule.getUserId();
+
+		User user = _userLocalService.getUser(userId);
+
+		if (StringUtil.equals(
+				objectValidationRule.getEngine(),
+				ObjectValidationRuleConstants.ENGINE_TYPE_COMPOSITE_KEY) &&
+			(keyObjectValidationRuleSettings.size() < 2)) {
+
+			throw new ObjectValidationRuleSettingCountException(
+				StringBundler.concat(
+					"The ", objectValidationRule.getName(user.getLocale()),
+					" Unique Composite Key must have at least two Object ",
+					"Fields to compose the Object Validation Rule"));
+		}
+
+		long objectValidationRuleId =
+			objectValidationRule.getObjectValidationRuleId();
+
+		_deleteOldObjectValidationRuleSettings(
+			objectValidationRuleId,
+			ObjectValidationRuleSettingConstants.
+				NAME_COMPOSITE_KEY_OBJECT_FIELD_ID,
+			keyObjectValidationRuleSettings,
+			objectValidationRuleSettingPersistence.findByOVRI_N(
+				objectValidationRuleId,
+				ObjectValidationRuleSettingConstants.
+					NAME_COMPOSITE_KEY_OBJECT_FIELD_ID));
+
+		_deleteOldObjectValidationRuleSettings(
+			objectValidationRuleId,
+			ObjectValidationRuleSettingConstants.NAME_OUTPUT_OBJECT_FIELD_ID,
+			outputObjectValidationRuleSettings,
+			objectValidationRuleSettingPersistence.findByOVRI_N(
+				objectValidationRuleId,
+				ObjectValidationRuleSettingConstants.
+					NAME_OUTPUT_OBJECT_FIELD_ID));
+
+		List<ObjectValidationRuleSetting> updatedObjectValidationRuleSettings =
+			new ArrayList<>();
+
+		_addOrUpdateObjectValidationRuleSetting(
+			objectValidationRuleId,
+			ObjectValidationRuleSettingConstants.
+				NAME_COMPOSITE_KEY_OBJECT_FIELD_ID,
+			keyObjectValidationRuleSettings,
+			updatedObjectValidationRuleSettings, userId);
+
+		_addOrUpdateObjectValidationRuleSetting(
+			objectValidationRuleId,
+			ObjectValidationRuleSettingConstants.NAME_OUTPUT_OBJECT_FIELD_ID,
+			outputObjectValidationRuleSettings,
+			updatedObjectValidationRuleSettings, userId);
+
```

Comment:
I believe isn't need perform the **addOrUpdate** phase twice based on the _name_.
In the **delete** phase this differentiation make sense, but here it seems unnecessary.
Since we have the list of news _objectValidationRuleSettings_, it would be better perform the loop placed into **_addOrUpdateObjectValidationRuleSetting** just once, using this list.